test/jasmine/assets/domain_ref/components.js

```diff
@@ -0,0 +1,793 @@
+    }
+    return axref;
+}
+
+// shape, annotation and image all take x0, y0, xref, yref, color parameters
+// x0, y0 are numerical values, xref, yref are strings that could be passed to
+// the xref field of an ANO (e.g., 'x2 domain' or 'paper'), color should be
+// specified using the 'rgb(r, g, b)' syntax
+// arotype can be 'shape', 'annotation', or 'image'
+// shapes take type=[line|rect], x1, y1
+// annotations take ax, ay, axref, ayref, (text is just set to "A" and xanchor
+// and yanchor are always set to left because these are text attributes which we
+// don't test)
+// images take xsize, ysize, xanchor, yanchor (sizing is set to stretch for simplicity
+// in computing the bounding box and source is something predetermined)
+function aroFromParams(arotype, x0, y0, xref, yref, color, opts) {
+    // fill with common values
+    var aro = {
+        xref: xref,
+        yref: yref
+    };
+    switch(arotype) {
+        case 'shape':
+            aro.x0 = x0;
+            aro.y0 = y0;
+            aro.x1 = opts.x1;
+            aro.y1 = opts.y1;
+            aro.type = opts.type;
+            aro.line = {
+                color: color
+            };
+            break;
+        case 'annotation':
+            aro.x = x0;
+            aro.y = y0;
+            aro.text = 'A';
+            aro.ax = opts.ax;
+            aro.ay = opts.ay;
+            aro.axref = opts.axref;
+            aro.ayref = opts.ayref;
+            aro.showarrow = true;
+            aro.arrowhead = 0;
+            aro.arrowcolor = color;
+            break;
+        case 'image':
+            aro.x = x0;
+            aro.y = y0;
+            aro.sizex = opts.sizex;
+            aro.sizey = opts.sizey;
+            aro.xanchor = opts.xanchor;
+            aro.yanchor = opts.yanchor;
+            aro.sizing = 'stretch';
+            aro.source = testImage;
+            break;
+        default:
+            throw 'Bad arotype: ' + arotype;
+    }
+    return aro;
+}
+
+// Calculate the ax value of an annotation given a particular desired scaling K
+// This also works with log axes by taking logs of each part of the sum, so that
+// the length in pixels is multiplied by the scalar
+function annaxscale(ac, c0) {
+    var ret;
+    ret = c0 + 2 * (ac - c0);
+    return ret;
+}
+
+// This tests to see that an annotation was drawn correctly.
+// Determinining the length of the arrow seems complicated due to the
+// rectangle containing the text, so we draw 2 annotations, one K times the
+// length of the other, and solve for the desired arrow length from the
+// length measured on the screen. This works because multiplying the length
+// of the arrow doesn't change where the arrow meets the text box.
+// xaxistype can be linear|log, only used if xref has type 'range' or 'domain',
+// same for yaxistype and yref
+function annotationTest(gd, layout, x0, y0, ax, ay, xref, yref, axref, ayref,
+    xaxistype, yaxistype, xid, yid) {
```

Comment:
Non-blocking: 
When having several input arguments to a function like this
```js
function annotationTest(gd, layout, x0, y0, ax, ay, xref, yref, axref, ayref,
    xaxistype, yaxistype, xid, yid) {
```
There is a chance to use wrong order in the caller(s).
It is possible to refactor it to use an object for last arguments. 
Something like:
```js
function annotationTest(gd, layout, opts) {
  var x0 = opts.x0;
  ...
```
Then change the caller to be 
```
annotationTest(gd, layout, {
    x0: x0, 
    y0: y0, 
    ax: ax,
    ay: ay, 
    xref: xref, 
    yref: yref, 
    axref: axref,
    ayref: ayref
    xaxistype: xaxistype, 
    yaxistype: yaxistype, 
    xid: xid, 
    yid: yid
});
```